union_pooling/union_pooling/PoolingRegion.py

```diff
@@ -0,0 +1,367 @@
+
+
+
+def _getPoolerClass(name):
+    if name=="union":
+      return UnionPooler
+    else:
+      raise RuntimeError("Invalid pooling implementation %s. Valid ones are: union" % (name))
+
+
+def _getDefaultPoolerClass():
+    return UnionPooler
+
+
+def _buildArgs(poolerClass, self=None, kwargs={}):
+  """
+  Get the default arguments from the function and assign as instance vars.
+
+  Return a list of 3-tuples with (name, description, defaultValue) for each
+    argument to the function.
+
+  Assigns all arguments to the function as instance variables of PoolingRegion.
+  If the argument was not provided, uses the default value.
+
+  Pops any values from kwargs that go to the function.
+
+  """
+  # Get the name, description, and default value for each argument
+  argTuples = getArgumentDescriptions(poolerClass.__init__)
+  argTuples = argTuples[1:]  # Remove "self"
+
+  # Get the names of the parameters to our own constructor and remove them
+  init = PoolingRegion.__init__
+  ourArgNames = [t[0] for t in getArgumentDescriptions(init)]
+  # Also remove a few other names that aren't in our constructor but are
+  #  computed automatically
+  ourArgNames += [
+    "inputDimensions",
+  ]
+  for argTuple in argTuples[:]:
+    if argTuple[0] in ourArgNames:
+      argTuples.remove(argTuple)
+
+  # Build the dictionary of arguments
+  if self:
+    for argTuple in argTuples:
+      argName = argTuple[0]
+      if argName in kwargs:
+        # Argument was provided
+        argValue = kwargs.pop(argName)
+      else:
+        # Argument was not provided; use the default value if there is one, and
+        #  raise an exception otherwise
+        if len(argTuple) == 2:
+          # No default value
+          raise TypeError("Must provide value for '%s'" % argName)
+        argValue = argTuple[2]
+      # Set as an instance variable if "self" was passed in
+      setattr(self, argName, argValue)
+
+  return argTuples
+
+
+def _getAdditionalSpecs(poolerClass=_getDefaultPoolerClass()):
+  """Build the additional specs in three groups (for the inspector)
+
+  Use the type of the default argument to set the Spec type, defaulting
+  to "Byte" for None and complex types
+
+  Determines the pooler parameters based on the selected implementation.
+  """
+  typeNames = {int: "UInt32", float: "Real32", str: "Byte", bool: "bool", tuple: "tuple"}
+
+  def getArgType(arg):
+    t = typeNames.get(type(arg), "Byte")
+    count = 0 if t == "Byte" else 1
+    if t == "tuple":
+      t = typeNames.get(type(arg[0]), "Byte")
+      count = len(arg)
+    if t == "bool":
+      t = "UInt32"
+    return (t, count)
+
+  def getConstraints(arg):
+    t = typeNames.get(type(arg), "Byte")
+    if t == "Byte":
+      return "multiple"
+    elif t == "bool":
+      return "bool"
+    else:
+      return ""
+
+  # Get arguments from pooler constructors, figure out types of
+  # variables and populate poolerSpec.
+  pArgTuples = _buildArgs(poolerClass)
```

Comment:
If I explicitly define poolerSpec, then I would have to modify this section every time we make a change to UnionPooler's __init__ arguments.  Right now it gets all the arguments that the UnionPooler has by calling _buildArgs. So, if you make a change to UnionPooler, it will make that change here as well automatically.